### DISCUSSION TITLE
tutorials/export/exporting_pcks

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#13] [User-contributed notes]
# tutorials/export/exporting_pcks

Use cases: Oftentimes one would like to add functionality to one's game after it has been deployed. Examples of this include... Downloadable Content: the ability to add features and content to one'...

https://docs.godotengine.org/en/latest/tutorials/export/exporting_pcks.html

**Comment by SuperTemich2005** [comment]
What do we think about updating the C# snippet to account for loading the DLL file?
it would be something like
```cs
private void YourFunction()
{
    Assembly.LoadFile("mod.dll")
    // This could fail if, for example, mod.pck cannot be found.
    var success = ProjectSettings.LoadResourcePack("res://mod.pck");

    if (success)
    {
        // Now one can use the assets as if they had them in the project from the start.
        var importedScene = (PackedScene)ResourceLoader.Load("res://mod_scene.tscn");
    }
}
``` 
i guess?

**Reply by horseyhorsey** [reply]
Do you not have to use? `Assembly.LoadFile(Path.Combine(Directory.GetCurrentDirectory(), "UnnamedProject.dll"));`. Was complaining about absolute paths when I used it.

**Reply by dangerz** [reply]
I'm still getting the error "Cannot instance script because the associated class could not be found".  It successfully loads my .dll and .pck file though.  Anyone able to overcome that?

**Comment by eth0net** [comment]
Are there any examples on building a modding tool as suggested here? I'd like to offer mod support as I learn to make games with godot and examples here would be a big help! :3

**Reply by Calinou** [reply]
See https://github.com/KoBeWi/Godot-Universal-Mod-Manager. If your game provides a modding API, the universal mod manager can make use of it (it will handle file loading).

**Reply by Awesomerly** [reply]
Someone I know made a modloader called GDWeave for Godot 3.5.2 and it lets people load/patch gdscript at runtime without the developer needing to add compatibility. Perhaps at some point it will get support for a more recent version.

**Comment by SememeS** [comment]
> No need for any of the path retrieval. In release, use "./" which starts relative from the binary

[reddit post link](https://www.reddit.com/r/godot/comments/zlk68u/comment/j07a384/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button)

**Comment by ag-radley** [comment]
This article seems to be a bit out of date. See: https://github.com/godotengine/godot/issues/75352

If you follow the code in the docs above when loading a .pck file and accompanying .dll containing C# scripts, you'll likely get an error saying the specified C# class can't be found. Basically you have to also call `Godot.Bridge.ScriptManagerBridge.LookupScriptsInAssembly` in addition to loading the .dll assembly in order to register it properly with Godot. It also may want an absolute path rather than a `res://` path (but both may work, I'm not sure).

Also, since no one seems to explain these things, the .dll you need to include with the .pck can be found in the `.godot\mono\temp\bin\ExportRelease\win-x64` of the exporting project's directory. The `\ExportDebug\win-x64` folder may also work, at least it did for me.

The code below worked for me.

```csharp
private static void LoadPckFileWithCSharpScripts()
{
    string dll_path = Path.Combine(Directory.GetCurrentDirectory(), "ExportingProject.dll");

    AssemblyLoadContext context = AssemblyLoadContext.GetLoadContext(typeof(Godot.Bridge.ScriptManagerBridge).Assembly);
    Assembly assembly = context.LoadFromAssemblyPath(dll_path); 
    Godot.Bridge.ScriptManagerBridge.LookupScriptsInAssembly(assembly);

    // This could fail if, for example, mod.pck cannot be found.
    bool success = ProjectSettings.LoadResourcePack("res://ExportingProject.pck");

    if (success)
    {
        // Now one can use the assets as if they had them in the project from the start.
        PackedScene imported_scene = (PackedScene)ResourceLoader.Load("res://SceneWithACSharpScriptAttached.tscn");
        Node instantiated_scene = imported_scene.Instantiate();
        Instance.AddChild(instantiated_scene);
        GD.Print($"LoadPckFileWithCSharpScripts: {instantiated_scene?.Name}");
    }
}
```

**Reply by chaojian-zhang** [reply]
Thanks! This is very helpful! I can confirm this works with Godot 4.4. The workflow looks like this:

1. Have some C# classes in a separate assembly (dll)/C# project;
2. In the main project, we somehow load the referenced assembly (in a number of plausible ways - using Assembly loader is just one way; We can also just reference the needed C# project in the main project)
3. Make sure to call `Godot.Bridge.ScriptManagerBridge.LookupScriptsInAssembly(assembly);` prior to load the resource pack
4. Load the resource pack corresponding to the C# assembly.

Now you can instantiate C# scripts in GDScript like any local C# scripts as documented!

**Reply by cgyfreedom-max** [reply]
@ag-radley Loading the DLL this way, using AOT compilation still requires loading a PCK package containing plaintext .cs code. Is there no way to avoid including plaintext code?

**Comment by yongan-game** [comment]
One thing I've found that's worth mentioning in the documentation is that, for web exports at least, you need to manually make the engine aware of the pack files to begin with.

For instance, if I've exported my music resources to `mus.pck` in the same directory as `index.html`, I'd need to put the following in the custom HTML shell, right above `engine.startGame()`:

```javascript
engine.preloadFile("mus.pck") // Extra line added here
setStatusMode('progress');
engine.startGame({ // ...
```

If you call `ProjectSettings.load_resource_pack()` without doing that, it will always return `false` from my experience.

**Comment by DanielKinsman** [comment]
I ran into a gotcha if you are doing the following:

* a browser game (web export)
* downloading extra pck at runtime (e.g. for separate levels) and loading them with `ProjectSettings.load_resource_pack`
* when the browser and webserver both support gzip compression (and not brotli)

You'll hit the following error:

```
ERROR: Condition "err != 0 && err != 1" is true. Returning: FAILED index.js:488:18
   at: _process (core/io/stream_peer_gzip.cpp:118) index.js:488:18
ERROR: url https://testserver/leveltutorial0000_web.pck failed 200 index.js:488:18
   at: push_error (core/variant/variant_utility.cpp:1024) index.js:488:18
ERROR: Couldn't download level tutorial0000 index.js:488:18
   at: push_error (core/variant/variant_utility.cpp:1024)
```

I think this is because the webbrowser is doing the gzip decompression, and then godot is also attempting to decompress the already decompressed bytes, but I'm not sure.

The solution is to disable gzip on the godot side and let the browser handle it by setting `HTTPRequest.accept_gzip=false`

**Comment by DexterFstone** [comment]
Are all *.pck files the same across different platforms? For example, if I exported a *.pck for Android, can I use it on Windows or the other way around? Does the Export Options affect the outcome? For instance, `architecture.`

**Reply by DanielKinsman** [reply]
The pcks will usually have different contents for each platform, e.g. textures in mobile will be in a different format than desktop.

**Reply by Calinou** [reply]
To clarify: while PCKs are not guaranteed to be platform-independent, they will be if your project doesn't use any VRAM-compressed textures. Only VRAM-compressed textures use platform-specific formats (S3TC/BPTC on desktop, ETC2/ASTC on mobile).

Basis Universal is another form of VRAM compression that *is* platform-independent, as it performs transcoding to the platform's preferred format at runtime.

**Comment by 1Googol** [comment]
Using Godot4.5. When export ZIP/PCK, `.tscn` became `.tscn.remap`, `.tres` became `.tres.remap`, this way I can't use `load("res://*.tscn")`. What should I do to load

**Reply by 1Googol** [reply]
solved it, in my case

```bash
var path = "res://example.tres"
if ResourceLoader.exists(path):
    var frames := load(path) as SpriteFrames
    sprite.sprite_frames = frames
```
